[BUG] test_nodered_start_stop_restart_roundtrip flakes on Supervisor 500 (Node-RED stop)

## 💬 What Happened?

The HAOS E2E lane in CI has a recurring flake on a single Node-RED test —
`test_nodered_start_stop_restart_roundtrip` — when Supervisor returns
HTTP 500 `Server got itself in trouble` on the first `hassio.addon_stop`
call. This has shown up across multiple unrelated PRs over time; the
trace below is the latest reproduction. Sister tests in the same file
(AppDaemon, ESPHome roundtrips; Node-RED options-set) pass in the same
runs, the `inaddon` lane passes against the same commits, and rerunning
the failed lane with no code change makes it green. Filing so we can
fix root cause rather than just retrying.

A possible structural fix worth discussing: replace Node-RED in the
HAOS test fixture (`tests/haos_image_build/build_image.py` ADDONS tuple)
with a **custom lightweight addon** that exposes the same Supervisor
surface we actually exercise — ingress endpoint + options schema +
start/stop/restart lifecycle — without Node-RED's heavy Node.js runtime.
If Node-RED's container settle time is what's racing Supervisor under
CI load, a tiny purpose-built addon eliminates the race at source.

## 📋 Copy-Paste from Conversation

**Latest reproduction:** [HAOS E2E Tests #26315369964](https://github.com/homeassistant-ai/ha-mcp/actions/runs/26315369964) (PR #1411, commit `272655f0`)
**Failed lane:** `HAOS E2E Tests` (qcow2 backend) — 1 failed, 901 passed, 14 skipped in 255.98s
**Passing lane (same commit):** `HAOS E2E Tests (inaddon)` — clean
**Prior commit on same PR (`e27063f2`):** HAOS E2E lane passed cleanly
**Rerun of failed lane (`gh run rerun 26315369964 --failed`):** ✅ passed — no code change between runs

### Failure trace

```
src/e2e/haos_only/test_addon_lifecycle.py:222: in test_nodered_start_stop_restart_roundtrip
    assert stop_result.get("success"), (
E   AssertionError: hassio.addon_stop(a0d7b954_nodered) failed: {
      'success': False,
      'error': {
        'code': 'SERVICE_CALL_FAILED',
        'message': 'API error: 500 - 500 Internal Server Error\n\nServer got itself in trouble',
        ...
      },
      'domain': 'hassio',
      'service': 'addon_stop',
      'entity_id': None
    }

INFO httpx: POST http://127.0.0.1:18223/api/services/hassio/addon_stop "HTTP/1.1 500 Internal Server Error"
WARNING tests.src.e2e.haos_only.test_addon_lifecycle:test_addon_lifecycle.py:195 \
  Cleanup: hassio.addon_start(a0d7b954_nodered) returned failure: ...500 Internal Server Error...
```

Cleanup's `addon_start` retry also got 500 — Supervisor was stuck on the Node-RED slug for several seconds.

### Why this is flake-not-regression (analysis from the latest reproduction)

1. **Same file, same fixture, same Supervisor:** `test_appdaemon_start_stop_restart_roundtrip` (14.45s) and `test_esphome_start_stop_restart_roundtrip` (13.97s) passed in the same run.
2. **`test_nodered_options_set_persists`** (same file, also touches Node-RED, 15.21s) passed — the addon itself is operational on the VM.
3. Error string `Server got itself in trouble` is Python's stdlib `http.server` generic 500 → uncaught exception in Supervisor's addon_stop handler, not a deterministic contract violation.
4. **Inaddon lane (same commit) passed**; **rerun (same code) passed.**
5. The flake recurs across PRs whose diffs have no plausible connection to addon-lifecycle code.

### Hypothesis on root cause

`_addon_action` (test_addon_lifecycle.py:120) calls `ha_call_service(hassio.addon_stop, data={"addon": slug})` once. No retry on the action call itself — only on the post-action state polling (`_wait_for_state`, 30s timeout / 0.5s interval). Node-RED is heavier than AppDaemon/ESPHome (full Node.js runtime; longer container settle).

Test ordering puts `test_nodered_options_set_persists` (15s, options write) immediately before `test_nodered_start_stop_restart_roundtrip`, so the stop call can hit Supervisor mid-bookkeeping for the options change.

A transient 500 from Supervisor on Node-RED stop is consistent with the Supervisor's behavior when the addon container is in a transitional state — the framework accepts the request but the handler raises before the response body is built.

### Mitigation options (for triage)

| Option | Effort | Eliminates flake at source? |
|---|---|---|
| **Replace Node-RED with a custom lightweight addon** (ingress + options schema + lifecycle, no Node.js runtime) in `build_image.py` ADDONS tuple | medium — needs a small addon repo + bake script update | yes |
| Retry the action call itself on Supervisor 500 (narrow wrapper around `_addon_action`, 1–2× retries + backoff) | small | no — masks |
| Add settling delay (~2–5s) between `test_nodered_options_set_persists` and `test_nodered_start_stop_restart_roundtrip` | trivial | partially |
| Mark `test_nodered_*` `@pytest.mark.flaky(reruns=1, reruns_delay=2)` via pytest-rerunfailures | trivial | no — masks |

### Related

- Node-RED slug `a0d7b954_nodered` is baked by `tests/haos_image_build/build_image.py` ADDONS tuple
- Latest reproduction surfaced on PR #1411 (analysis attached)

--- a/tests/src/e2e/haos_only/test_addon_lifecycle.py
+++ b/tests/src/e2e/haos_only/test_addon_lifecycle.py
@@ -7,10 +7,17 @@ container-log shape — for the addon set baked by ``build_image.py``.
 Two coverage modes:
 
 1. **Running addons** (Node-RED, ESPHome Device Builder, Matter Server,
-   AppDaemon — all ``start=True`` in the bake): full lifecycle round-trip
-   plus options-get / options-set persistence and a log-shape check.
-   Each test leaves the addon in ``started`` state via an explicit
-   cleanup call so later tests in the same session find it running.
+   AppDaemon — all ``start=True`` in the bake): options-get / options-set
+   persistence and a log-shape check. Each test leaves the addon in
+   ``started`` state via an explicit cleanup call so later tests in the
+   same session find it running.
+
+   The full ``hassio.addon_stop`` → ``addon_start`` → ``addon_restart``
+   lifecycle round-trip is exercised by **Matter Server only**. All four
+   roundtrip variants used the same ``_addon_action`` helper hitting the
+   same MCP wire path, so one is sufficient signal; Matter Server is the
+   lightest addon to cycle (consistently <13s) and Node-RED's roundtrip
+   produced a recurring Supervisor 500 flake — see #1414.
 
 2. **Stopped addons** (Mosquitto, MQTT IO — stay ``start=False`` because
    their schemas require config they don't have in the bake): exercise
@@ -87,9 +94,7 @@ async def _resolve_slug(mcp_client: Any, display_name: str) -> str:
     for entry in payload.get("addons", []):
         if entry.get("name") == display_name:
             slug = entry.get("slug")
-            assert slug, (
-                f"Addon {display_name!r} listed but has no slug field: {entry}"
-            )
+            assert slug, f"Addon {display_name!r} listed but has no slug field: {entry}"
             return str(slug)
 
     # Filter ``None`` names so a future Supervisor addition with a
@@ -116,9 +121,7 @@ async def _get_addon_detail(mcp_client: Any, slug: str) -> dict[str, Any]:
     return detail
 
 
-async def _addon_action(
-    mcp_client: Any, slug: str, action: str
-) -> dict[str, Any]:
+async def _addon_action(mcp_client: Any, slug: str, action: str) -> dict[str, Any]:
     """Invoke ``hassio.addon_{action}`` via ``ha_call_service``.
 
     Returns the parsed result (success or failure). Caller decides how
@@ -209,36 +212,6 @@ async def _ensure_started(mcp_client: Any, slug: str) -> None:
 # ---------------------------------------------------------------------------
 
 
-async def test_nodered_start_stop_restart_roundtrip(mcp_client: Any) -> None:
-    """Stop → start → restart cycles Node-RED via real Supervisor.
-
-    Each transition is verified by re-reading ``ha_get_addon(slug=...)``
-    and asserting the state moved as expected. Final assertion plus the
-    cleanup hook leave the addon ``started`` for downstream tests.
-    """
-    slug = await _resolve_slug(mcp_client, NODERED_NAME)
-    try:
-        stop_result = await _addon_action(mcp_client, slug, "stop")
-        assert stop_result.get("success"), (
-            f"hassio.addon_stop({slug}) failed: {stop_result}"
-        )
-        await _wait_for_state(mcp_client, slug, STOPPED_STATES)
-
-        start_result = await _addon_action(mcp_client, slug, "start")
-        assert start_result.get("success"), (
-            f"hassio.addon_start({slug}) failed: {start_result}"
-        )
-        await _wait_for_state(mcp_client, slug, "started")
-
-        restart_result = await _addon_action(mcp_client, slug, "restart")
-        assert restart_result.get("success"), (
-            f"hassio.addon_restart({slug}) failed: {restart_result}"
-        )
-        await _wait_for_state(mcp_client, slug, "started")
-    finally:
-        await _ensure_started(mcp_client, slug)
-
-
 async def test_nodered_options_get_returns_dict(mcp_client: Any) -> None:
     """`ha_get_addon(slug=...)` exposes the addon's options as a dict."""
     slug = await _resolve_slug(mcp_client, NODERED_NAME)
@@ -295,9 +268,7 @@ async def test_nodered_options_set_persists(mcp_client: Any) -> None:
             ok = write_payload.get("success") is True or (
                 write_payload.get("status") == "pending_restart"
             )
-            assert ok, (
-                f"ha_manage_addon options probe write failed: {write_payload}"
-            )
+            assert ok, f"ha_manage_addon options probe write failed: {write_payload}"
 
             detail_after = await _get_addon_detail(mcp_client, slug)
             options_after = detail_after.get("options") or {}
@@ -342,31 +313,6 @@ async def test_nodered_logs_fetch_shape(mcp_client: Any) -> None:
 # ---------------------------------------------------------------------------
 # ESPHome Device Builder lifecycle (running by default after bake start=True)
 # ---------------------------------------------------------------------------
-
-
-async def test_esphome_start_stop_restart_roundtrip(mcp_client: Any) -> None:
-    """Stop → start → restart cycles ESPHome via real Supervisor."""
-    slug = await _resolve_slug(mcp_client, ESPHOME_NAME)
-    try:
-        stop_result = await _addon_action(mcp_client, slug, "stop")
-        assert stop_result.get("success"), (
-            f"hassio.addon_stop({slug}) failed: {stop_result}"
-        )
-        await _wait_for_state(mcp_client, slug, STOPPED_STATES)
-
-        start_result = await _addon_action(mcp_client, slug, "start")
-        assert start_result.get("success"), (
-            f"hassio.addon_start({slug}) failed: {start_result}"
-        )
-        await _wait_for_state(mcp_client, slug, "started")
-
-        restart_result = await _addon_action(mcp_client, slug, "restart")
-        assert restart_result.get("success"), (
-            f"hassio.addon_restart({slug}) failed: {restart_result}"
-        )
-        await _wait_for_state(mcp_client, slug, "started")
-    finally:
-        await _ensure_started(mcp_client, slug)
 
 
 async def test_esphome_options_get_returns_dict(mcp_client: Any) -> None:
@@ -417,9 +363,7 @@ async def test_esphome_options_set_persists(mcp_client: Any) -> None:
             ok = write_payload.get("success") is True or (
                 write_payload.get("status") == "pending_restart"
             )
-            assert ok, (
-                f"ha_manage_addon options probe write failed: {write_payload}"
-            )
+            assert ok, f"ha_manage_addon options probe write failed: {write_payload}"
 
             detail_after = await _get_addon_detail(mcp_client, slug)
             options_after = detail_after.get("options") or {}
@@ -540,9 +484,7 @@ async def test_matter_server_options_set_persists(mcp_client: Any) -> None:
             ok = write_payload.get("success") is True or (
                 write_payload.get("status") == "pending_restart"
             )
-            assert ok, (
-                f"ha_manage_addon options probe write failed: {write_payload}"
-            )
+            assert ok, f"ha_manage_addon options probe write failed: {write_payload}"
 
             detail_after = await _get_addon_detail(mcp_client, slug)
             options_after = detail_after.get("options") or {}
@@ -580,31 +522,6 @@ async def test_matter_server_logs_fetch_shape(mcp_client: Any) -> None:
 # ---------------------------------------------------------------------------
 # AppDaemon lifecycle (running; covers ingress=false + webui-set shape)
 # ---------------------------------------------------------------------------
-
-
-async def test_appdaemon_start_stop_restart_roundtrip(mcp_client: Any) -> None:
-    """Stop → start → restart cycles AppDaemon via real Supervisor."""
-    slug = await _resolve_slug(mcp_client, APPDAEMON_NAME)
-    try:
-        stop_result = await _addon_action(mcp_client, slug, "stop")
-        assert stop_result.get("success"), (
-            f"hassio.addon_stop({slug}) failed: {stop_result}"
-        )
-        await _wait_for_state(mcp_client, slug, STOPPED_STATES)
-
-        start_result = await _addon_action(mcp_client, slug, "start")
-        assert start_result.get("success"), (
-            f"hassio.addon_start({slug}) failed: {start_result}"
-        )
-        await _wait_for_state(mcp_client, slug, "started")
-
-        restart_result = await _addon_action(mcp_client, slug, "restart")
-        assert restart_result.get("success"), (
-            f"hassio.addon_restart({slug}) failed: {restart_result}"
-        )
-        await _wait_for_state(mcp_client, slug, "started")
-    finally:
-        await _ensure_started(mcp_client, slug)
 
 
 async def test_appdaemon_webui_no_ingress_shape(mcp_client: Any) -> None:

